### ISSUE TITLE
Optimize Lua onSort callback

### ISSUE THREAD

**Issue by smichel17** [mpcjanssen/simpletask-android#931]
`onSort` is very important to my current setup — it allows me to have tasks with `(A)` priority and tasks due today at the top of my list, followed by lower-priority or and less urgent tasks. Unfortunately, it's also very slow on large lists, and waiting maybe half a second every time I complete or modify a task gets old *really* fast.

I'm going to spend some time soon looking into optimizing the lua callbacks. Not sure exactly when since I'm kinda busy atm, but likely sooner than later since it's impacting my productivity. I think right now we're calling `onSort` once per comparison; probably we just need to cache the result so we only call it once per task. I don't think it's a trivial change, but shouldn't be *too* hard.

I'll do #727 at the same time.

**Comment by mpcjanssen** [comment]
I am not sure if caching onSort will give much benefit. Any reasonable sorting algorithm will only sort two tasks against each other once.

**Comment by smichel17** [comment]
Yeah, I was thinking of adding something to the `Task` object like

```
val luaSortString by lazy {
    Interpreter.onSortCallback(luaModule, this)
}
```

But of course that won't work because a task doesn't have any conceptualization of the query it's in, plus the value needs to change when the active query changes.

Might end up making a wrapper for `Task` (eg, `FilteredTask`, or just caching the result for sorting specifically, though I'm not sure exactly what the best way to do that is since right now we're just setting up the comparator. Maybe `Interpreter` should be caching the results? I dunno, I need to think about it.

**Comment by smichel17** [comment]
> `val str = Interpreter.onSortCallback(moduleName, it)`

In my 200+ task list, this is being called 3000+ times. I'm doing a fair bit of work in my `onSort` -> slow.

**Comment by mpcjanssen** [comment]
I would do the caching in the comparator where it should be imo. Anything else will introduce tight coupling.

**Comment by mpcjanssen** [comment]
A simple hashmap with task and result would be enough.

**Comment by smichel17** [comment]
I just wanted to check whether the onSort callback is used anywhere else that we'd like to cache. But I think for now you're probably right about caching in the comparator, even if only to avoid opening the possibility of cache invalidation bugs.

**Comment by mpcjanssen** [comment]
> There are only two hard things in Computer Science: cache invalidation, naming things , and off-by-1 errors.
